File: C++/InsertionSort/InsertionSort.cpp
#include "InsertionSort.h"

//sort dataset in ascending order (without using key)
int InsertionSort::SortAscending(int arr[], int size)
{
	int counter = 0;

	for (int i = 1; i < size; i++) 
	{
		counter++;
		int j = i;

		while (j > 0 && arr[j] < arr[j - 1]) 
		{
			int temp = arr[j];
			arr[j] = arr[j - 1];
			arr[j - 1] = temp;
			j--;
		}
	}
	return counter;
}

//sort dataset in descending order (without using key)
int InsertionSort::SortDescending(int arr[], int size)
{
	int counter = 0;

	for (int i = 1; i < size; i++)
	{
		counter++;
		int j = i;

		while (j > 0 && arr[j] > arr[j - 1])
		{
			int temp = arr[j];
			arr[j] = arr[j - 1];
			arr[j - 1] = temp;
			j--;
		}
	}
	return counter;
}

//sort dataset in ascending order (using key)
int InsertionSort::SortAscendingKeyMethod(int arr[], int size)
{
	int counter = 0;

	for (int i = 1; i < size; i++)
	{
		counter++;
		int key = arr[i];
		int j = i - 1;
		while (j >= 0 && key < arr[j])
		{
			arr[j + 1] = arr[j];
			j--;
		}
		arr[j + 1] = key;
	}
	return counter;
}

//sort dataset in descending order (using key)
int InsertionSort::SortDescendingKeyMethod(int arr[], int size)
{
	int counter = 0;

	for (int i = 1; i < size; i++)
	{
		counter++;
		int key = arr[i];
		int j = i - 1;
		while (j >= 0 && key > arr[j])
		{
			arr[j + 1] = arr[j];
			j--;
		}
		arr[j + 1] = key;
	}
	return counter;
}

File: C#/InsertionSort/InsertionSort.cs
using System.Reflection;
using System.Security.Permissions;

namespace Algorithms
{
    public class InsertionSort
    {
        //insertion sort without using a key for comparison (ascending order)
        public (int[], int) SortAscending(int[] array)
        {
            int counter = 0;

            for (int i = 1; i < array.Length; i++)
            {
                counter++;
                int j = i;

                //whilst the focused number is less than number to its right
                //swap numbers and continue until not true
                while (j > 0 && array[j] < array[j - 1]) 
                {
                    int temp = array[j];
                    array[j] = array[j - 1];
                    array[j - 1] = temp;
                    j--;
                }
            }
            return (array, counter);
        }

        //insertion sort without using a key for comparison (descending order)
        public (int[], int) SortDescending(int[] array)
        {
            int counter = 0;

            for (int i = 1; i < array.Length; i++)
            {
                counter++;
                int j = i;

                //whilst the focused number is less than number to its left
                //swap numbers and continue until not true
                while (j > 0 && array[j] > array[j - 1])
                {
                    int temp = array[j];
                    array[j] = array[j - 1];
                    array[j - 1] = temp;
                    j--;
                }
            }
            return (array, counter);
        }

        //insertion sort using a key for comparison (ascending order)
        public (int[], int) SortKeyMethodAscending(int[] array)
        {
            int counter = 0;
            for (int i = 1; i < array.Length; i++)
            {
                counter++;
                int key = array[i];
                int j = i - 1;
                while (j >= 0 && key < array[j])
                {
                    array[j + 1] = array[j];
                    j--;
                }
                array[j + 1] = key;
            }
            return (array, counter);
        }

        //insertion sort using a key for comparison (descending order)
        public (int[], int) SortKeyMethodDescending(int[] array)
        {
            int counter = 0;
            for (int i = 1; i < array.Length; i++)
            {
                counter++;
                int key = array[i];
                int j = i - 1;
                while (j >= 0 && key > array[j])
                {
                    array[j + 1] = array[j];
                    j--;
                }
                array[j + 1] = key;
            }
            return (array, counter);
        }
    }
}
